File: PAT_Basic_Level_by_go/1088 三人行/1088 三人行.go
// 题目：1088 三人行
// 实现原理：甲是两位数（10~99），乙是甲的数字反转。已知：|甲 - 乙| / x = 丙，
// 且乙 = 丙 * y。从 99 往下枚举甲的取值，找到满足条件的最大甲。然后根据
// 甲、乙、丙的值与 M 的大小关系，分别输出 "Cong"（大于）、"Ping"（等于）
// 或 "Gai"（小于）。

package main

import (
	"bufio"
	"fmt"
	"os"
)

// main 是程序入口，读取 M、x、y，找出满足条件的甲并输出比较结果。
func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var m, x, y int
	fmt.Fscan(in, &m, &x, &y)

	// 甲: 2-digit number. 乙: digits reversed. 丙 = |甲-乙|/x, also 乙 = 丙*y.
	// So 乙 = |甲-乙|/x * y  =>  |甲-乙| = 乙*x/y. Need integer 丙.
	bestJia := -1
	bestYi := -1
	bestBing := -1
	found := false
	// 从大到小枚举甲，保证找到的是最大值
	for jia := 99; jia >= 10; jia-- {
		yi := (jia%10)*10 + jia/10 // 乙 = 甲的数字反转
		diff := jia - yi
		if diff < 0 {
			diff = -diff
		}
		// 丙 = diff / x must be integer, and yi == bing*y
		if x == 0 {
			continue
		}
		if diff%x != 0 { // 丙必须是整数
			continue
		}
		bing := diff / x
		if bing <= 0 {
			continue
		}
		if yi != bing*y { // 验证乙 = 丙 * y
			continue
		}
		bestJia = jia
		bestYi = yi
		bestBing = bing
		found = true
		break
	}
	if !found {
		fmt.Fprintln(out, "No Solution")
		return
	}
	// rel 比较 v 与 m 的大小关系，返回对应字符串
	rel := func(v int) string {
		if v > m {
			return "Cong"
		}
		if v == m {
			return "Ping"
		}
		return "Gai"
	}
	fmt.Fprintf(out, "%d %s %s %s\n", bestJia, rel(bestJia), rel(bestYi), rel(bestBing))
}

File: PAT_Basic_Level_by_cpp/1088 三人行/1088 三人行.cpp
// 1088 三人行
//
// 实现原理：
// 甲的能力值是 2 位正整数；把甲的 2 个数字调换位置得到乙的能力值；
// 甲乙两人能力差是丙的能力值的 X 倍；乙的能力值是丙的 Y 倍。
// 即：乙 = 甲的数字调换，丙 = 乙 / Y，且 |甲 - 乙| = 丙 * X。
// 若解不唯一，取甲的最大解；若解不存在，输出 "No Solution"。
//
// 关键步骤：
// 1. 从大到小遍历所有 2 位正整数作为甲（保证取最大解）。
// 2. 计算乙 = 甲的数字调换；若乙能被 Y 整除，则丙 = 乙 / Y。
// 3. 检查 |甲 - 乙| == 丙 * X 是否成立，成立则找到解。
// 4. 输出甲的能力值，以及甲、乙、丙三人与"我"（M）的关系：
//    比 M 强输出 Cong，相等输出 Ping，比 M 弱输出 Gai。
//
// 复杂度分析：
// 时间复杂度：O(1)，最多遍历 90 个 2 位数。
// 空间复杂度：O(1)，仅使用常数个变量。
#include <iostream>
#include <cmath>

using namespace std;

int main() {
    int m, x, y;
    if (!(cin >> m >> x >> y)) return 0;

    int jia = 0, yi = 0, bing = 0;
    bool found = false;
    for (int a = 99; a >= 10; --a) {          // 从大到小遍历，保证取甲的最大解
        int b = (a % 10) * 10 + a / 10;       // 乙 = 甲的数字调换
        if (b % y != 0) continue;             // 乙必须是 Y 的整数倍
        int c = b / y;                        // 丙 = 乙 / Y
        if (abs(a - b) == c * x) {            // 甲乙差 = 丙 * X
            jia = a; yi = b; bing = c;
            found = true;
            break;
        }
    }

    if (!found) { cout << "No Solution" << endl; return 0; }

    cout << jia;
    int vals[3] = {jia, yi, bing};
    for (int i = 0; i < 3; ++i) {
        cout << ' ';
        if (vals[i] > m) cout << "Cong";
        else if (vals[i] == m) cout << "Ping";
        else cout << "Gai";
    }
    cout << endl;
    return 0;
}
